docs/design/draft/initial-design.md

```diff
@@ -0,0 +1,379 @@
+backing up and restoring OpenShift Persistent Volumes, and it forms
+one component of a larger backup and restore strategy. One of the main
+motivations for the Data Mover is that it combines the minimum
+downtime of snapshot-based backups with the restore flexibility of
+filesystem-based strategies. If the user will always restore to the
+same volume type (and, where relevent, to the same region), then
+velero backup and restore using the Velero Plugin for CSI will handle
+it. If the limited downtime of snapshot-based backup isn't needed,
+then Velero backup and restore via using restic copy will work. The
+Data Mover combines the advantages of both approaches. In addition, it
+will allow an alternate Data Mover plugin to be used instead of the
+Restic plugin to meet additional user requirements.
+
+## Data Mover as related to the Velero CSI Plugin
+
+Velero (and the Velero CSI plugin) will not be a requirement to use
+the Data Mover. However, creating a Velero backup with CSI
+snapshotting enabled is one way to provide the per-PVC CSI snapshots
+required to use the Data Mover.  The Velero backup of the snapshot
+won't be used. Alternatively, the user can manually create the
+VolumeSnapshot resource or make use of some other infrastructure
+automation tools to generate the snapshot.
+
+Here is a summary of the actions taken by the Velero CSI plugin on backup:
+
+- The Velero `VCBackupItemAction` creates the `VolumeSnapshot`. The
+  (Kubernetes) Snapshot controller (along with the csi-snapshotter)
+  then creates `VolumeSnapshotContent`.
+- The Velero `VolumeSnapshotBackupItemAction` annotates the
+  `VolumeSnapshot`, includes the associated `VolumeSnapshotContent`
+  and `VolumeSnapshotClass` as `additionalItems` in the backup.
+- the Velero `VolumeSnapshotContentBackupItemAction` and
+  `VolumeSnapshotClassBackupItemAction` include relevant secrets as
+  `additionalItems` in the backup
+
+## Data Mover CRDs
+
+The following CRDs will be needed by the Data Mover:
+
+## `DataMoverStorageLocation`
+
+We need a way to model a storage location for PVC storage. The
+simplest thing to do here would be just to reuse either Velero's
+`BackupStorageLocation` model, or CAM's `MigStorage` model. We could
+define our own with its own model, or we could even have a more
+flexible system where a `DataMoverStorageLocation` could optionally
+reference a Velero `BackupStorageLocation`, a CAM `MigStorage`, or a
+third data-mover-internal model. Regardless of the choice here, the
+metadata needs are approximately the same as the pre-existing Velero
+and CAM CRDs. In subsequent examples, `StorageLocation` is used as a
+stand-in for whatever is decided on here.
+
+### `DataMoverBackup`
+
+The `DataMoverBackup` is the resource created to initiate a Backup
+operation in the Data Mover controller. The main things that this must
+contain are a list of PVCs to back up and a `StorageLocation`
+reference.
+
```

Comment:
I have a small doubt here (apologies if it is a trivial question). 

But who creates the this "DataMoverBackup" resource and what is the trigger point? I am trying to understand the operation from the perspective of a backup application which performs the following operations.

1) Issues CSI snapshots
2) Has its own custom data mover (based on this design) which moves the data to a target location.

In this situation, once the CSI snapshots have been created by the backup application, who creates the resource "DataMoverBackup"?

Reply:
This actually brings up a good question. One part which isn't quite as clear to me as it should be is exactly how the Data Mover will be invoked. This is how I had envisioned the DMB/DMR resources fitting into this:
1) Some existing tool generates CSI snapshots: This could be velero. It could be manually creating a VolumeSnapshot. It could be some other backup/restore infrastructure tool
2) Whatever invokes the DataMover creates a DataMoverBackup (could be a UI we build, could be some other tool via API, could be manual)
3) in the process of acting on the DataMoverBackup, the DataMover controller creates a Deployment which contains the relevant DataMover plugin (including the appropriate binary for whatever 3rd party/custom data mover tool is invoked). For our reference implementation, this means this deployment will need our restic plugin as well as the restic binary.

@jwmatthews Does the above sound right? Other than the question of deciding how the DataMover and the DataMover plugins will interact, the question of what invokes the DataMover in the first place (i.e. how we create the DataMoverBackup/DataMoverRestore for the DataMover to operate on) is one that we really haven't defined (or even talked much about) yet. This is analogous to the question for CAM of "how does the MigPlan/MigMIgration get created?

Reply:
My understanding of data mover is that, the "datamover controller" would be watching for CSI snapshots. And once it finds a new CSI snapshot created, the data mover would start its work of backing up of the contents of the snapshot.

This is how I understand things.

1) With Velero invoking the CSI snapshots.

  - Velero invokes CSI snapshots as part of the backup operation.
  - Data mover would notice the creation of new CSI snapshot (as it would be watching for it), and would start backing up 
    of the contents of the snapshot to the target.
  - Velero would backup the kubernetes resources.

2) A backup application invoking the CSI snapshots.

  - The backup application invokes the CSI snapshots.
  - Data mover would notice the creation of new CSI snapshot (as it would be watching for it), and would start backing up 
    of the contents of the snapshot to the target.

NOTE: The data mover would be a custom data mover plugin written by the backup (both  of the above 2 cases) which would be based on the data mover design we would recommend.

This is my understanding. My 1st comment (https://github.com/konveyor/data-mover/pull/1#discussion_r437470231) was also made with respect to this understanding.

Reply:
So what you describe is what I reference at the bottom of the doc as an alternate implementation. It really depends on whether the requirements are to automatically drive this off CSI snapshots or whether there is some value in a higher-level grouping of PVCs together as a "backup" of some sort. The latter was my assumption throughout, but I mention the former as another possibility. We really need to figure out (at a high level/requirements point of view) which we're dealing with before doing much more, since it's a fairly fundamental difference in how this will all work.

Once we know this we can figure out what happens on restore. In the explicit backup case, the restore will be as-described in this document right now. If the data mover is operating on individual CSI snapshots as they appear, then we'll have to revisit the restore side as well. It could be that the DataMoverRestore as described would still be valid, but instead of referencing a DataMoverBackup, it would reference individual volumes.  What's less clear is how we'd identify them -- by CSI snapshot ID, or by PVC name/namespace (then grabbing the most recent snapshotted version), etc.